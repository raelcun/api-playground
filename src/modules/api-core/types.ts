import * as t from 'io-ts'
import Koa from 'koa'

import { RolesV } from '@lib/rbac'

interface KoaRequest<T> extends Koa.Request {
  body: T
}

export interface KoaContext<T> extends Koa.Context {
  request: KoaRequest<T>
}
export type KoaContext2<T> = Koa.ParameterizedContext<Koa.DefaultState, { request: { body: T } }>

type a = Koa.Middleware<Koa.DefaultState, { request: { body: T } }>

export type Middleware<T> = (ctx: KoaContext<T>, next: () => Promise<void>) => Promise<void>

export const tokenV = t.type({
  userId: t.string,
  role: RolesV,
})
export type Token = t.TypeOf<typeof tokenV>
