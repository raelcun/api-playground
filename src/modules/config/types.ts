import * as t from 'io-ts'

import { LoggingConfig } from '@lib/logger'
import { DeepPartial, NonEmptyString } from '@lib/utils'

export interface FullConfig {
  isProduction: boolean
  env: string
  server: {
    jwtSecret: string
    securePort: number
    insecurePort: number
    enableSSL: boolean
  }
  logging: LoggingConfig
  aws: {
    region: string
    credentials: {
      accessKeyId: string
      secretAccessKey: string
    }
  }
}

export type PartialConfig<T> = DeepPartial<T>

export type ConfigProvider = () => FullConfig

export const envVarsV = t.type({
  NODE_ENV: NonEmptyString,
  JWT_SECRET: NonEmptyString,
  AWS_ACCESS_KEY_ID: NonEmptyString,
  AWS_SECRET_ACCESS_KEY: NonEmptyString,
})
export type EnvVars = t.TypeOf<typeof envVarsV>
