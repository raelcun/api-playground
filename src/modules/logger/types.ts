import { Logger as PinoLogger } from 'pino'

export type LogMethods = 'fatal' | 'trace' | 'error' | 'info' | 'debug' | 'warn'

export type Logger = Pick<PinoLogger, LogMethods | 'child'>

export type LoggerFactory = () => Logger
