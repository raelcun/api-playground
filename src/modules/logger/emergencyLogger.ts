import pino, { Logger } from 'pino'

export const getEmergencyLogger = (): Logger =>
  pino({
    name: 'emergency',
    level: 'trace',
  })
