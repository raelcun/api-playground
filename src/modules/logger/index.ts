export * from './logger'
export { middleware as logRequest } from './logRequestMiddleware'
