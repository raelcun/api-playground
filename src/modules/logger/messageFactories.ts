import { Context } from 'koa'

import { Message } from './types'

export const getMessageFromContext = (ctx: Context): Message => ({
  request: {
    url: ctx.href,
    headers: ctx.headers,
    method: ctx.method,
  },
  response: {
    status: ctx.response.status,
    headers: ctx.response.headers,
    body: ctx.response.body,
  },
})

export const createTimingMessageSegment = (
  startNanos: bigint,
  startTime: Date,
  timingType: string,
): Message['timing'] => {
  const duration = Number(process.hrtime.bigint() - startNanos) / 1e6
  const startTimestamp = startTime.toISOString()
  const endTimestamp = new Date(startTime.getTime() + duration).toISOString()

  return {
    type: timingType,
    start: startTimestamp,
    end: endTimestamp,
    duration: Number(process.hrtime.bigint() - startNanos) / 1e6,
  }
}
