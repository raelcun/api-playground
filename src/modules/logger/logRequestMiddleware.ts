import { Middleware } from 'koa'

import { getSystemLogger } from './logger'
import { LoggerFactory } from './types'

const hrTimeToMs = (hrTime: [number, number]) => hrTime[0] * 1e3 + hrTime[1] / 1e6

export const createMiddleware = (createLogger: LoggerFactory): Middleware => async (ctx, next) => {
  const start = process.hrtime()

  let err: any = undefined
  try {
    await next()
  } catch (e) {
    err = e
  }

  createLogger().info({
    method: ctx.method,
    path: ctx.url,
    responseStatus: ctx.status,
    duration: hrTimeToMs(process.hrtime(start)),
  })

  if (err !== undefined) throw err
}

export const middleware = createMiddleware(getSystemLogger)
