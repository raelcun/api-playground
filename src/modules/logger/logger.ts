import { createLogger, Logger } from '@lib/logger'
import { getConfig } from '@modules/config'

export const getLogger = createLogger(() => getConfig().logging)

export const getSystemLogger = (): Logger => getLogger('system')
