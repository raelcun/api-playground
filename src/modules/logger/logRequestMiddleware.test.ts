import { Lazy } from 'fp-ts/lib/function'
import { Next } from 'koa'

import { Logger } from '@modules/logger/types'
import { createKoaContext, createMockNext, installStaticClock } from '@root/utils'

import { createMiddleware } from './logRequestMiddleware'

const createMockLogger = (): Logger => ({
  info: jest.fn(),
  fatal: jest.fn(),
  trace: jest.fn(),
  error: jest.fn(),
  debug: jest.fn(),
  warn: jest.fn(),
  child: jest.fn(),
})

const eatError = async <T>(fn: Lazy<Promise<T>>) => {
  try {
    await fn()
  } catch {}
}

const createMockMiddleware = () => {
  const mockLogger = createMockLogger()
  const mockInfo = mockLogger.info as ReturnType<typeof jest.fn>
  const mockContext = createKoaContext({ url: 'MOCK_URL', method: 'GET' })
  const middleware = createMiddleware(() => mockLogger)

  return {
    mockInfo,
    mockContext,
    callMiddleware: (next: Next) => middleware(mockContext, next),
  }
}

describe('logRequestMiddleware', () => {
  const getClock = installStaticClock()

  // let mockInfo: ReturnType<typeof jest.fn>
  // let callMiddleware: (mockNext: Next) => Promise<void>

  // beforeEach(async () => {
  //   const mockLogger = createMockLogger()
  //   mockInfo = mockLogger.info as ReturnType<typeof jest.fn>
  //   const mockContext = createKoaContext({ url: 'MOCK_URL', method: 'GET' })

  //   callMiddleware = mockNext => createMiddleware(() => mockLogger)(mockContext, mockNext)
  // })

  test('log should match snapshot when next throws', async () => {
    const { callMiddleware, mockInfo, mockContext } = createMockMiddleware()
    const mockNext = createMockNext(() => {
      getClock().tick(1)
      mockContext.status = 500
      throw new Error()
    })
    await eatError(callMiddleware(mockNext))
    expect(mockInfo.mock.calls[0][0]).toMatchSnapshot()
  })

  test('log should match snapshot when next succeeds', async () => {
    const { callMiddleware, mockInfo, mockContext } = createMockMiddleware()
    const mockNext = createMockNext(() => {
      getClock().tick(1)
      mockContext.status = 200
      throw new Error()
    })
    await eatError(() => callMiddleware(mockNext))
    expect(mockInfo.mock.calls[0][0]).toMatchSnapshot()
  })

  test('should pass through error', async () => {
    expect.assertions(1)

    const { callMiddleware } = createMockMiddleware()
    const mockNext = createMockNext(() => {
      throw new Error('foobar')
    })

    try {
      await callMiddleware(mockNext)
    } catch (e) {
      expect(e.message).toEqual('foobar')
    }
  })
})
