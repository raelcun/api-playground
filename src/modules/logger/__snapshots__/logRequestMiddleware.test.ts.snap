// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logRequestMiddleware log should match snapshot when next succeeds 1`] = `
Object {
  "duration": 1,
  "method": "GET",
  "path": "http://test.com/MOCK_URL",
  "responseStatus": 200,
}
`;

exports[`logRequestMiddleware log should match snapshot when next throws 1`] = `
Object {
  "duration": 1,
  "method": "GET",
  "path": "http://test.com/MOCK_URL",
  "responseStatus": 500,
}
`;
